src/onionreq/builder.cpp

```diff
@@ -43,6 +44,92 @@ EncryptType parse_enc_type(std::string_view enc_type) {
+}
+
+template <>
+void Builder::set_destination(SnodeDestination destination) {
+    destination_x25519_public_key.reset();
+    ed25519_public_key_.reset();
+    destination_x25519_public_key.emplace(destination.node.x25519_pubkey);
+    ed25519_public_key_.emplace(destination.node.ed25519_pubkey);
+}
+
+template <>
+void Builder::set_destination(ServerDestination destination) {
+    destination_x25519_public_key.reset();
+
+    host_.emplace(destination.host);
+    target_.emplace("/oxen/v4/lsrpc");  // All servers support V4 onion requests
+    endpoint_.emplace(destination.endpoint);
+    protocol_.emplace(destination.protocol);
+    method_.emplace(destination.method);
+
+    if (destination.port)
+        port_.emplace(*destination.port);
+
+    if (destination.headers)
+        headers_.emplace(*destination.headers);
+
+    if (destination.query_params)
+        headers_.emplace(*destination.query_params);
+
+    destination_x25519_public_key.emplace(destination.x25519_pubkey);
+}
+
+ustring Builder::generate_payload(std::optional<ustring> body) const {
+    // If we don't have the data required for a server request, then assume it's targeting a
+    // service node and, therefore, the `body` is the payload
+    if (!host_ || !target_ || !endpoint_ || !protocol_ || !method_ ||
+        !destination_x25519_public_key)
+        return body.value_or(ustring{});
+
+    // Otherwise generate the payload for a server request
+    auto headers_json = nlohmann::json::object();
+
+    if (headers_)
+        for (const auto& [key, value] : *headers_) {
+            // Some platforms might automatically add this header, but we don't want to include it
+            if (key != "User-Agent")
+                headers_json[key] = value;
+        }
+
+    if (body && !headers_json.contains("Content-Type"))
+        headers_json["Content-Type"] = "application/json";
+
+    // Need to ensure the endpoint has a leading forward slash so add it if it's missing
+    auto endpoint = *endpoint_;
+
+    if (!endpoint.empty() && endpoint.front() != '/')
+        endpoint = '/' + endpoint;
+
+    // If we have query parameters, add them to the endpoint to be included in the payload
+    if (query_params_ && !query_params_->empty()) {
+        std::string query_string = "";
+
+        for (auto& query : *query_params_)
+            query_string += query.first + "=" + query.second + "&";
+
+        // Drop the extra '&' from the end
+        endpoint += "?" + query_string.substr(0, query_string.size() - 1);
+    }
+
+    // Structure the request information
+    nlohmann::json request_info{
+            {"method", *method_}, {"endpoint", endpoint}, {"headers", headers_json}};
+    auto request_info_dump = request_info.dump();
+    std::vector<std::string> payload{request_info_dump};
+
+    // If we were given a body, add it to the payload
+    if (body.has_value())
+        payload.emplace_back(std::string(from_unsigned(body->data()), body->size()));
```

Comment:
There's a `ustring_view` version of `from_unsigned` called `from_unsigned_sv` that makes this simpler:
```suggestion
        payload.emplace_back(from_unsigned_sv(*body));
```

Reply:
Oh nice - I didn't realise that it would automatically convert the `std::string_view` to a `std::string` for the `std::vector<std::string> payload` 👍

Reply:
Have made this change